### ISSUE TITLE
[asyncio] `asyncio.gather()` not propagating CancellationError properly on 3.9 and 3.10

### ISSUE THREAD

**Issue by Tinche** [python/cpython#112534]
# Bug report

### Bug description:

I found a small regression in `asyncio.gather` while working on a library of mine.

As far as I can bisect, the change appeared in 3.9.11 (3.9.10 is good) and is still present in the latest versions of 3.9 and 3.10. My test suite would indicate the problem is **not** present on 3.11 and 3.12.

A minimal reproducer:

```python
from asyncio import CancelledError, current_task, gather, get_running_loop, run, sleep


async def main():
    this = current_task()
    get_running_loop().call_later(0.01, lambda: this.cancel(msg="TEST"))

    try:
        await gather(sleep(10), sleep(10))
    except CancelledError as exc:
        print(repr(exc))


run(main())
```

The correct output is:
```
CancelledError('TEST')
```

The output on 3.9.11+ is:

```
CancelledError()
```

The issue is that the cancel message is not propagated correctly when the parent task is cancelled.

The cancel message seems to get erased in the process of handling everything. I couldn't figure out where exactly.

### CPython versions tested on:

3.9, 3.12

### Operating systems tested on:

macOS

**Comment by gvanrossum** [comment]
Thanks for the bug report. As you may know, cancellation propagation is not always done correctly. Perhaps you can find the problem by bisecting? You already have a good range -- it was introduced between 3.9.10 and 3.9.11.

**Comment by zware** [comment]
3.9 and 3.10 are both in security-fix-only mode, so if 3.11+ are fine there's nothing to be done here.
